### PR TITLE
[REVIEW] Graph sage example

### PR DESCRIPTION
**Pull request by VibhuJawa**
This PR depends upon https://github.com/rapidsai/cugraph/pull/2924 . 

We add `Graph-Sage` example to claim `dgl` , `cugraph` integration. 

Waiting on SWIPAT before merging

CC: @TristonC

### CODE REVIEW

**Review by tingyu66 — CHANGES_REQUESTED** [review]
Looks good. Just have the following comments:

- I don't think we need to support `mode` other than `cugraph_storage`, and the code have assumed to use cugraph_storage in a few places, indicated by calling `graph.get_node_storage()`
- Need more explanation around the batch_size and the rmm option to users.

**Review by wangxiaoyunNV — APPROVED** [review]

File: python/cugraph-dgl/examples/graphsage/node-classification.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    parser.add_argument("--use_rmm", action="store_true")
+    args = parser.parse_args()
+    if not torch.cuda.is_available():
+        args.mode = "cpu"
+    print(f"Training in {args.mode} mode.")
+
+    # load and preprocess dataset
+    print("Loading data")
+    dataset = AsNodePredDataset(DglNodePropPredDataset("ogbn-products"))
+    g = dataset.g
```
[the remark this answers is not in the document]
Comment (wangxiaoyunNV) [reply python/cugraph-dgl/examples/graphsage/node-classification.py:202]: agree with Tingyu, `g = dataset[0]` per DGL API

**Review by wangxiaoyunNV — COMMENTED** [review]
see my comments in the code

**Review by wangxiaoyunNV — CHANGES_REQUESTED** [review]
Please see me comments in the code.

File: python/cugraph-dgl/examples/graphsage/node-classification.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    else:
+        g = g.to("cuda" if args.mode == "puregpu" else "cpu")
+    device = torch.device("cpu" if args.mode == "cpu" else "cuda")
+
+    # create GraphSAGE model
+    feat_shape = (
+        g.get_node_storage(key="feat", ntype="_N")
+        .fetch(torch.LongTensor([0]).to(device), device=device)
+        .shape[1]
+    )
```
[the remark this answers is not in the document]
Comment (wangxiaoyunNV) [reply python/cugraph-dgl/examples/graphsage/node-classification.py:223]: why do not you first get the `shape[1]`, then use `to(device)` ? this will save some data movement time.

File: python/cugraph-dgl/examples/graphsage/node-classification.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    # create sampler & dataloader
+    train_idx = dataset.train_idx.to(device)
+    val_idx = dataset.val_idx.to(device)
+    sampler = NeighborSampler(
+        [10, 10, 10],  # fanout for [layer-0, layer-1, layer-2]
+        prefetch_node_feats=["feat"],
+        prefetch_labels=["label"],
+    )
+    use_uva = args.mode == "mixed"
+    batch_size = 1024 * 20
```
[the remark this answers is not in the document]
Comment (wangxiaoyunNV) [reply python/cugraph-dgl/examples/graphsage/node-classification.py:127]: will the loss function converge with batch_size = 20k?

File: python/cugraph-dgl/examples/graphsage/node-classification.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph-dgl/examples/graphsage/node-classification.py:202]: Thanks, will change this

File: python/cugraph-dgl/examples/graphsage/node-classification.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph-dgl/examples/graphsage/node-classification.py:223]: Yes, this snippet works across all DGL storage devices  following the rfc at [link](https://hackmd.io/6PJixJj7Ssy19I-N9CV1Ag#Extending-DGL-Graph-Sampling-Pipeline ) 

Also, see this example [link](https://github.com/dmlc/dgl/blob/2caef17eb8481ea42503a9843aad77af2710b425/python/dgl/storages/pytorch_tensor.py#LL28C47-L28C47)

> why do not you first get the shape[1], then use to(device) ? this will save some data movement time.

1. The `to(device)` call here is for the indices argument which needs the indices to be on the same device as the feature. 

2. We also are only moving 1 single index to device here which is an array of size 1 so i dont think this is expensive at all.

File: python/cugraph-dgl/examples/graphsage/node-classification.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+                h = F.relu(h)
+                h = self.dropout(h)
+        return h
+
+    def inference(self, g, device, batch_size):
+        """Conduct layer-wise inference to get all the node embeddings."""
+        all_node_ids = torch.arange(0, g.num_nodes()).to(device)
+        feat = g.get_node_storage(key="feat", ntype="_N").fetch(
+            all_node_ids, device=device
+        )
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph-dgl/examples/graphsage/node-classification.py:54]: Yes

File: python/cugraph-dgl/examples/graphsage/node-classification.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+def layerwise_infer(device, graph, nid, model, batch_size):
+    model.eval()
+    with torch.no_grad():
+        pred = model.inference(graph, device, batch_size)  # pred in buffer_device
+        pred = pred[nid]
+
+        label = (
+            graph.get_node_storage(key="label")
+            .fetch(nid.to(device), device=device)
+            .to(pred.device)
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph-dgl/examples/graphsage/node-classification.py:111]: Yes, it should work across storage devices

File: python/cugraph-dgl/examples/graphsage/node-classification.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph-dgl/examples/graphsage/node-classification.py:127]: > will the loss function converge with batch_size = 20k?

We currently get `0.697` as the test accuracy but on the default training epochs compared to `0.7632`  with a smaller batch size. I have not done hyper parameter tuning which will be needed to see how much perf we can get with this batch size. 

I did not focus on accuracy on this workflow but this is more of an integration test to ensure we actually work as expected with cugraph backend

File: python/cugraph-dgl/examples/graphsage/node-classification.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+        default="cugraph_storage",
+        choices=["cpu", "mixed", "puregpu", "cugraph_storage"],
+        help="Training mode."
+        " 'cpu' for CPU training,"
+        " 'mixed' for CPU-GPU mixed training, "
+        " 'puregpu' for CPU-GPU mixed training, "
+        " 'cugraph_storage' for pure-GPU training.",
+    )
+
+    parser.add_argument("--use_rmm", action="store_true")
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph-dgl/examples/graphsage/node-classification.py:193]: Thanks , will do

File: python/cugraph-dgl/examples/graphsage/node-classification.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+    with torch.no_grad():
+        pred = model.inference(graph, device, batch_size)  # pred in buffer_device
+        pred = pred[nid]
+
+        label = (
+            graph.get_node_storage(key="label")
+            .fetch(nid.to(device), device=device)
+            .to(pred.device)
+        )
+        # label = graph.ndata['label'][nid].to(pred.device)
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph-dgl/examples/graphsage/node-classification.py:113]: Will do

**Review by alexbarghi-nv — DISMISSED** [review]
Looks like Xiaoyun addressed everything that needed additional review.  Approved.

File: python/cugraph-dgl/examples/graphsage/node-classification.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tingyu66) [reply python/cugraph-dgl/examples/graphsage/node-classification.py:202]: I see. That's indeed an awkward design.

[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph-dgl/examples/graphsage/node-classification.py]: Resolved with https://github.com/rapidsai/cugraph/pull/2925/commits/cac6dba904b216de636c79a258d8effc07197e10

File: python/cugraph-dgl/examples/graphsage/node-classification.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph-dgl/examples/graphsage/node-classification.py]: Done . Added it

**Comment by VibhuJawa** [comment]
@tingyu66 , @wangxiaoyunNV , All changes have been addressed or commented on, Please let me know if there is anything else we need to do.

**Review by tingyu66 — DISMISSED** [review]
LGTM. Thanks!

**Review by wangxiaoyunNV — DISMISSED** [review]

**Comment by kingmesal** [comment]
@VibhuJawa any status updates?

**Review by rlratzel — CHANGES_REQUESTED** [review]
LGTM, just a couple of minor comments

File: python/cugraph-dgl/examples/graphsage/node-classification.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+import torch.nn.functional as F
+import torchmetrics.functional as MF
+import dgl.nn as dglnn
+from dgl.data import AsNodePredDataset
+from dgl.dataloading import DataLoader, NeighborSampler, MultiLayerFullNeighborSampler
+from ogb.nodeproppred import DglNodePropPredDataset
+import tqdm
+import argparse
+
+# force creating cugraph context
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph-dgl/examples/graphsage/node-classification.py]: Removed

[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph-dgl/examples/graphsage/node-classification.py]: I switched the import order so this is no longer needed.

**Comment by VibhuJawa** [comment]
@rlratzel , All your reviews have been addressed.

**Review by rlratzel — APPROVED** [review]

**Comment by VibhuJawa** [comment]
@tingyu66 , @alexbarghi-nv Waiting on you all's approval too.

**Review by alexbarghi-nv — APPROVED** [review]
👍

**Review by tingyu66 — APPROVED** [review]
:+1:
